### ISSUE TITLE
build fails with `error: expected initializer before ‘:’ token', and others

### ISSUE THREAD

**Issue by kortschak** [PacificBiosciences/blasr#91]
I am failing to build blasr. It looks to me like it may be a gcc version issue, but the *only* information I can see anywhere about compiler version requirements is [here](https://github.com/PacificBiosciences/blasr/issues/28#issuecomment-49918188) which says 4.4.3 is fine (though that is an old comment).

```
$ git rev-parse HEAD
d8733e29ea421a171f70e66624cb153b41bba3e1
$ (cd libcpp && git rev-parse HEAD)
e2acb33134be5bacec6e8b0e47484b73c980f128
$ gcc --version
gcc (GCC) 4.4.7 20120313 (Red Hat 4.4.7-4)
Copyright (C) 2010 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

```
$ make build-submodule
# DON'T use pbbam which is not on github.
cd libcpp && NOPBBAM=true HDF5_LIB=/usr/local/lib/libhdf5.so HDF5_INCLUDE=/usr/local/include ./configure.py
NOT1
make -C libcpp
make[1]: Entering directory `/home/kortschak/blasr/libcpp'
make -C pbdata all
make[2]: Entering directory `/home/kortschak/blasr/libcpp/pbdata'
g++ -std=c++11 -pedantic -MMD -MP    -I. -Imatrix -Ireads -Iqvs -Imetagenome -Isaf -Iutils -Ialignment -MF"CCSSequence.d" -MT"CCSSequence.o CCSSequence.d" -c CCSSequence.cpp -o CCSSequence.o
cc1plus: error: unrecognized command line option "-std=c++11"
make[2]: *** [CCSSequence.o] Error 1
make[2]: Leaving directory `/home/kortschak/blasr/libcpp/pbdata'
make[1]: *** [all] Error 2
make[1]: Leaving directory `/home/kortschak/blasr/libcpp'
make: *** [build-submodule] Error 2
$ make blasr
g++ -O3 -g -std=c++0x -pedantic -Wall -Wuninitialized -Wno-div-by-zero -MMD -MP -w -fpermissive -fno-builtin-malloc -fno-builtin-calloc -fno-builtin-realloc -fno-builtin-free -fno-omit-frame-pointer  -I/home/kortschak/blasr/libcpp/alignment -I/home/kortschak/blasr/libcpp/pbdata -I/home/kortschak/blasr/libcpp/hdf -I/usr/local/include  -c -o Blasr.o Blasr.cpp
In file included from MappingParameters.h:18,
                 from MappingIPC.h:6,
                 from Blasr.cpp:15:
/home/kortschak/blasr/libcpp/alignment/datastructures/alignment/FilterCriteria.hpp:42: error: expected ‘;’ before ‘float’
Blasr.cpp: In function ‘std::vector<T_AlignmentCandidate*, std::allocator<T_AlignmentCandidate*> > SelectAlignmentsToPrint(std::vector<T_AlignmentCandidate*, std::allocator<T_AlignmentCandidate*> >, MappingParameters&, const int&)’:
Blasr.cpp:2658: error: expected initializer before ‘:’ token
Blasr.cpp:4464: error: expected primary-expression at end of input
Blasr.cpp:4464: error: expected ‘;’ at end of input
Blasr.cpp:4464: error: expected primary-expression at end of input
Blasr.cpp:4464: error: expected ‘)’ at end of input
Blasr.cpp:4464: error: expected statement at end of input
Blasr.cpp:4464: error: expected ‘}’ at end of input
make: *** [Blasr.o] Error 1
```

**Comment by pb-cdunn** [comment]
[pbdagcon](https://github.com/PacificBiosciences/pbdagcon) requires gcc-4.8.1+, so that will probably be our baseline going forward. We need to note that for blasr. Will it be possible for you to upgrade?

The good news is that clang should work, but we can't say that reliably until we have CI tests, which we can't have until we acquire installable binaries for pbbam and hdf5. But I'll add the compiler note today.

**Comment by pb-cdunn** [comment]
**blasr_libcpp** still compiles fine with `-std=c++0x`. (I just tried it, on my own branch.) That means, with a tiny change to the makefiles, it could probably still compile with some older gcc versions, [maybe even gcc-4.4.3](-std=c++0x).

How important is this to you? If very, we could probably give you an unsupported branch which uses `-std=c++0x` and still works with older compilers. We really do need to upgrade to C++11 at some point, so we could not offer support or continual updates on that branch. But it would work. Is that what you prefer? Or is it possible for you to upgrade your compiler?

**Comment by kortschak** [comment]
I've been trying to get gcc 4.8.1 and this is not a trivial exercise when trying to avoid building gcc from source (which I would like if at all possible).

I have asked our IT support to look into making RH  devtoolset available, which is what is necessary to get 4.8.1 for RHEL6 but I have not heard back from them yet.

Out of curiosity, what is the drive for moving to C++11?

**Comment by pb-cdunn** [comment]
> Out of curiosity, what is the drive for moving to C++11?

Lots of reasons. First, C++0x *is* C++11, just not the final version. With the former, you have to keep track of which [C++11 features are missing](https://gcc.gnu.org/projects/cxx0x.html). In my mind, the [improved memory model](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2007/n2429.htm) alone is worth the upgrade. I remember when Java's memory model used to cause huge threading problems, until Java-5. With C++11, C++ actually has a *better* model than Java (according to experts at these things).

Basically, C++11 (and C++14, and soon C++17) makes C++ a more modern language. However, the difficulty in *installing* C++ compilers just shows another reason we really should be switching to an even *more modern* language, like Rust or Nim, or even Ocaml or Go. But we're heavily invested in C++.

If IT fails for you, let me know and I'll make a branch for you.

**Comment by kortschak** [comment]
Thanks for the explanation. I agree with your assessment, reduction of cognitive load is a definite win.

I have had a lot of success using Go for bioinformatic work (I have a whole genome aligner ported from C and now written in Go with reasonably comparable performance and improved safety). The simplicity of the language is a definite bonus when it comes to development and review.

I'll ping or close this when the compiler issue at my end is sorted out or not.

**Comment by kortschak** [comment]
OK. That has worked for me.

For others that follow and are using RHEL6, installing devtoolset-3-toolchain and then doing

```
xport CC=/opt/rh/devtoolset-3/root/usr/bin/gcc
export CPP=/opt/rh/devtoolset-3/root/usr/bin/cpp
export CXX=/opt/rh/devtoolset-3/root/usr/bin/c++
export LIBS=/opt/rh/devtoolset-3/root/usr/lib/gcc/x86_64-redhat-linux/4.9.1/libstdc++_nonshared.a
```

gives the environment that is needed for the build.

@pb-cdunn when you add the note for the minumum gcc version, you may also want to modify the text of README.INSTALL.md to add "then" after the "To make the 'libcpp' libraries: make build-submodule" step, since the current reading does not indicate clearly that it is a dep of the complete build (yes, it should be obvious, but ...).

**Comment by pb-cdunn** [comment]
Thanks for the info. And yes, Go is sweet. I'm curious: What aligner did you port?

**Comment by kortschak** [comment]
https://github.com/biogo/examples/tree/master/krishna is a port of Bob Edgar and Eugene Myers' PALS aligner.
